src/back/space/src/api/space.cpp

```diff
@@ -0,0 +1,164 @@
+				return Delete(req, res);
+			case server::http::HttpMethod::kHead:
+				return Head(req, res);
+			default:
+				throw std::runtime_error("Unsupported");
+				break;
+		}
+	} catch(const errors::Unauthorized& e) {
+		req.SetResponseStatus(server::http::HttpStatus::kUnauthorized);
+		res["err"] = e.what();
+		return res.ExtractValue();
+	} catch(const errors::BadRequest& e) {
+		req.SetResponseStatus(server::http::HttpStatus::kBadRequest);
+		res["err"] = e.what();
+		return res.ExtractValue();
+	} catch(const errors::NotFound& e) {
+		req.SetResponseStatus(server::http::HttpStatus::kNotFound);
+		res["err"] = e.what();
+		return res.ExtractValue();
+	} catch(const errors::Conflict& e) {
+		req.SetResponseStatus(server::http::HttpStatus::kConflict);
+		res["err"] = e.what();
+		return res.ExtractValue();
+	} catch(const std::exception& e) {
+		LOG_WARNING() << "Fail to process space handle with method: "
+			<< req.GetMethodStr() << " err: " << e.what();
+		res["err"] = "Fail to process space";
+		req.SetResponseStatus(server::http::HttpStatus::kInternalServerError);
+	}
+
+	return res.ExtractValue();
+}
+
+formats::json::Value Space::Get(
+	const server::http::HttpRequest& req,
+	formats::json::ValueBuilder& res) const
+{
+	if (req.HasArg("id"))
+	{
+		const auto userId = req.GetHeader(headers::kUserId);
+		if (userId.empty())
+			throw errors::Unauthorized{};
+
+		const auto id = parseUUID(req, "id");
+		res = _s.GetById(id, userId);
+	}
+	else if (req.HasArg("key"))
+	{
+		const auto userId = req.GetHeader(headers::kUserId);
+		if (userId.empty())
+			throw errors::Unauthorized{};
+
+		const auto key = req.GetArg("key");
+		if (!_s.IsValidUUID(key))
+			if (!_s.CheckKeyByRegex(key))
+				throw errors::BadRequest{"Key must be valid"};
+		res = _s.GetByKey(key, userId);
+	}
+	else if (req.HasArg("link"))
+	{
+		const auto linkId = parseUUID(req, "link");
+		res = _s.GetByLink(linkId);
+	}
+	else
+		throw errors::BadRequest{"No arguments"};
+
+	return res.ExtractValue();
+}
+
+formats::json::Value Space::Delete(
+	const server::http::HttpRequest& req,
+	formats::json::ValueBuilder& res) const
+{
+	const auto userId = req.GetHeader(headers::kUserId);
+	if (userId.empty())
+		throw errors::Unauthorized{};
+
+	const auto id = parseUUID(req, "id");
+	if (!_s.IsSpaceOwner(id, userId))
+		throw errors::NotFound();
+
+	_s.Delete(id);
+
+	return res.ExtractValue();
+}
+
+formats::json::Value Space::Post(
+	const server::http::HttpRequest& req,
+	const formats::json::Value& body,
+	formats::json::ValueBuilder& res) const
+{
+	const auto userId = req.GetHeader(headers::kUserId);
+	if (userId.empty())
+		throw errors::Unauthorized{};
+
+	auto space = body.As<model::Space>();
+
+	if (_s.isSpaceExistsByKey(space.key))
+		throw errors::Conflict{"Invalid key"};
+
+	if (!_s.IsKeyValid(space.key))
+		throw errors::BadRequest("Can't use that key");
+
+	if (!_s.KeyAdditionalCheck(space.key, userId))
+		throw errors::BadRequest("Can't use such key");
```

Comment:
Сначала нужно выполнить лёгкие проверки которые не требуют лезть в базу
```suggestion
	if (!_s.KeyCreateCheck(space.key))
		throw errors::BadRequest("Can't use that key");

	if (_s.isSpaceExistsByKey(space.key))
		throw errors::Conflict{"Invalid key"};
```

Reply:
Готово.